dev/src/lobster/flat_gltf.h

```diff
@@ -0,0 +1,789 @@
+
+  // turn on 6 attribute arrays
+  glEnableVertexAttribArray(0); // Position
+  glEnableVertexAttribArray(1); // Normal
+  glEnableVertexAttribArray(2); // UV
+  glEnableVertexAttribArray(3); // Color (not used)
+  glEnableVertexAttribArray(4); // Weights
+  glEnableVertexAttribArray(5); // Indices
+
+  // generate a vertex buffer and an index buffer
+  glGenBuffers(2, &mesh.vao[1]);
+
+  _VERBOUT_ << "Vertex buffer address: " << mesh.vao[1] << std::endl;
+  _VERBOUT_ << "Index buffer address:  " << mesh.vao[2] << std::endl;
+
+  // switch to the vertex buffer
+  glBindBuffer(GL_ARRAY_BUFFER, mesh.vao[1]);
+  // upload the vertices, passing the size in bytes
+  glBufferData(GL_ARRAY_BUFFER, vertices.size() * sizeof(Vertex), &vertices[0],
+               streaming ? GL_STREAM_DRAW : GL_STATIC_DRAW);
+  _VERBOUT_ << "Number of vertices: " << vertices.size() << std::endl;
+  _VERBOUT_ << "Size of Vertex:     " << sizeof(Vertex) << std::endl;
+
+  // perhaps an inconsistent use of data here but I'm not sure what to do
+  // sizeof returns a size_t which can be cast to anything since it's just a
+  // number
+  std::size_t pointer = 0;
+  glVertexAttribPointer(0, 4, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Position
+  pointer += sizeof(vertices[0].Position);
+  glVertexAttribPointer(1, 3, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Normal
+  pointer += sizeof(vertices[0].Normal);
+  glVertexAttribPointer(2, 2, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // UV
+  pointer += sizeof(vertices[0].UV);
+  glVertexAttribPointer(3, 4, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Color (not used)
+  pointer += sizeof(vertices[0].Color);
+  glVertexAttribPointer(4, 4, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Weights
+  pointer += sizeof(vertices[0].Weights);
+  glVertexAttribPointer(5, 4, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Indices
+
+  // switch to the index buffer
+  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, mesh.vao[2]);
+  // upload the indices, passing the size in bytes
+  glBufferData(GL_ELEMENT_ARRAY_BUFFER, indices.size() * sizeof(Index),
+               &indices[0], streaming ? GL_STREAM_DRAW : GL_STATIC_DRAW);
+  _VERBOUT_ << "Number of indices: " << indices.size() << std::endl;
+  _VERBOUT_ << "Size of Index:     " << sizeof(Index) << std::endl << std::endl;
+
+  // switch out of the VAO
+  glBindVertexArray(0);
+
+  // failure conditions are currently unimplemented
+  mesh.success = true;
+
+  // store the number of indices
+  mesh.indexCount = indices.size();
+
+  return mesh;
+}
+
+void DrawMesh(Mesh &mesh, float4x4 modelViewMat, float4x4 projMat) {
+  if (!mesh.success)
+    return;
+
+  float4x4 mvp = modelViewMat * projMat;
```

Comment:
snake_case please

Reply:
This is the kind of shit I don't like doing, which is why I think there should be [a plugin system](https://github.com/fluffrabbit/flat_gltf/blob/master/why_not_plugins.md).

Reply:
That would be nice, and is technically not hard: as you see the only function that really needs to be exposed to the Lobster core is the one that creates the bindings, and that could easily sit in a DLL. Then you could have a way to declare for a program which plugins it uses (since it is also needed at compile time).

It is a bit of work to structure things like that though, since Lobster builds on 6 platforms, not all of which have DLL mechanisms. So there would need to be a way to build it the old fashioned way as well. And I don't like having to depend on external repos for that.

Then there's the issue that some of these, like gltf, naturally share a lot with the rest of the "engine", so keeping that all separate would be a bit of duplication.

You're currently adding this to the core of the lobster engine, so it will have to conform to that engine. More work, but results in better code in the end. And I'll help maintain it from here on.

If you feel that having to bow to ugly conventions like snake_case makes it not worth contribution, you could have instead made a completely stand-alone gltf loading library in its own repo. Then you could have suggested it for use in Lobster by adding it to `dev/external/fluff_gltf` or whatever. The only things needing to go into the main code base would be the bindings. You could have kept your own coding style. On the downside, I would have still been picky on adopting it, as I only want well maintained high quality libraries in `external`, so more burden on you to prove that. And I wouldn't be maintaining any of it, so it had better be high quality at the moment I integrate it. Usually that means many github stars :P

Reply:
Had I known what the `external` folder was for, I would have gone that route. Ah, well. If you want to absorb this thing into your body I should be grateful and help the process go smoothly.

As for shared libraries, the issues you mention are part of the whole DLL flavor. It wouldn't work on mobile and there would be lots of code redundancy, but those 10 MB DLLs would be someone else's problem, not yours. Current core libs are fine I guess, but I found your choice of IMGUI over other functionality a little strange from a game context. You can only add so many libraries before Lobster becomes bloated.